File: packages/engine/src/cpus/6502/instructions/plp.js
import { CycleType } from '..'

/**
 * PulL Processor flags (from the stack)
 *
 * @example
 * MODE          SYNTAX        HEX LEN TIM
 * Implied       PLP           $28  1   4
 */
export function* plp() {
  this.fetch()
  if (--this.pc < 0) this.pc = 0xffff
  yield CycleType.GET

  if (++this.sp > 0xff) this.sp = 0
  yield CycleType.GET

  this.status = this.read(0x100 | this.sp)
  this.breakCommand = false
  yield CycleType.GET
}

File: packages/engine/src/cores/famibob/cpu/instructions/plp.zig
const std = @import("std");
const CPU = @import("../cpu.zig").CPU;
const Addressing = @import("../opcode.zig").Addressing;

// PulL Processor flags (from the stack)
//
// MODE          SYNTAX        HEX LEN TIM
// Implied       PLP           $28  1   4

pub fn plp(self: *CPU, addressing: Addressing) void {
    _ = addressing;
    if (self.cycle_counter == 2) {
        _ = self.fetch();
        self.pc -%= 1;
        self.next_cycle = .read;
        return;
    } else if (self.cycle_counter == 3) {
        self.sp +%= 1;
        self.next_cycle = .read;
        return;
    } else {
        self.setP(self.read(0x100 | @as(u16, self.sp)));
        self.b = false;
    }
}

test {
    _ = @import("plp.test.zig");
}
